### PR TITLE
Remove EHOSTUNREACH check from SSLHandshake

### LINKED ISSUE #904: Firestore crashing on Android 4 in offline mode
* Android Studio version: 3.5
  * Firebase Component: Firestore
  * Component version: 21.1.1

#### Steps to reproduce:

1. Add a snapshot listener to a collection/document reference
2. Launch the app on Android 4.x
3. Turn the flight mode on
4. The app crashes

The app works perfectly fine as long as the internet connection is available, so it should not be a TLS issue as suggested by the exception message.

#### Relevant Code:

```
10-12 21:31:55.423 3498-3575/com.fitifyworkouts.bodyweight.workoutapp.debug W/Firestore: (21.1.1) [OnlineStateTracker]: Could not reach Cloud Firestore backend. Connection failed 1 times. Most recent error: Status{code=UNAVAILABLE, description=null, cause=java.net.ConnectException: failed to connect to firestore.googleapis.com/2a00:1450:400e:809::200a (port 443): connect failed: EHOSTUNREACH (No route to host)
        at libcore.io.IoBridge.connect(IoBridge.java:114)
        at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:192)
        at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
        at java.net.Socket.startupSocket(Socket.java:567)
        at java.net.Socket.<init>(Socket.java:226)
        at javax.net.DefaultSocketFactory.createSocket(DefaultSocketFactory.java:51)
        at io.grpc.okhttp.OkHttpClientTransport$4.run(OkHttpClientTransport.java:541)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
        at java.lang.Thread.run(Thread.java:841)
     Caused by: libcore.io.ErrnoException: connect failed: EHOSTUNREACH (No route to host)
        at libcore.io.Posix.connect(Native Method)
        at libcore.io.BlockGuardOs.connect(BlockGuardOs.java:85)
        at libcore.io.IoBridge.connectErrno(IoBridge.java:127)
        at libcore.io.IoBridge.connect(IoBridge.java:112)
        at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:192) 
        at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172) 
        at java.net.Socket.startupSocket(Socket.java:567) 
        at java.net.Socket.<init>(Socket.java:226) 
        at javax.net.DefaultSocketFactory.createSocket(DefaultSocketFactory.java:51) 
        at io.grpc.okhttp.OkHttpClientTransport$4.run(OkHttpClientTransport.java:541) 
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587) 
        at java.lang.Thread.run(Thread.java:841) 
    }
    This typically indicates that your device does not have a healthy Internet connection at the moment. The client will operate in offline mode until it is able to successfully connect to the backend.
10-12 21:31:55.743 3498-3498/com.fitifyworkouts.bodyweight.workoutapp.debug D/AndroidRuntime: Shutting down VM
10-12 21:31:55.743 3498-3498/com.fitifyworkouts.bodyweight.workoutapp.debug W/dalvikvm: threadid=1: thread exiting with uncaught exception (group=0x9ccbbb20)
10-12 21:31:55.753 3498-3498/com.fitifyworkouts.bodyweight.workoutapp.debug E/main: java.lang.IllegalStateException: The Cloud Firestore client failed to establish a secure connection. This is likely a problem with your app, rather than with Cloud Firestore itself. See https://bit.ly/2XFpdma for instructions on how to enable TLS on Android 4.x devices.
        at com.google.firebase.firestore.remote.AbstractStream.close(com.google.firebase:firebase-firestore@@21.1.1:278)
        at com.google.firebase.firestore.remote.AbstractStream.handleServerClose(com.google.firebase:firebase-firestore@@21.1.1:387)
        at com.google.firebase.firestore.remote.AbstractStream$StreamObserver.lambda$onClose$3(com.google.firebase:firebase-firestore@@21.1.1:151)
        at com.google.firebase.firestore.remote.AbstractStream$StreamObserver$$Lambda$4.run(com.google.firebase:firebase-firestore@@21.1.1)
        at com.google.firebase.firestore.remote.AbstractStream$CloseGuardedRunner.run(com.google.firebase:firebase-firestore@@21.1.1:67)
        at com.google.firebase.firestore.remote.AbstractStream$StreamObserver.onClose(com.google.firebase:firebase-firestore@@21.1.1:137)
        at com.google.firebase.firestore.remote.FirestoreChannel$1.onClose(com.google.firebase:firebase-firestore@@21.1.1:131)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:700)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:399)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:500)
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:65)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:592)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$700(ClientCallImpl.java:508)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:632)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:152)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:265)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
        at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor$DelayedStartFactory.run(com.google.firebase:firebase-firestore@@21.1.1:224)
        at java.lang.Thread.run(Thread.java:841)
     Caused by: java.net.ConnectException: failed to connect to firestore.googleapis.com/2a00:1450:400e:809::200a (port 443): connect failed: EHOSTUNREACH (No route to host)
        at libcore.io.IoBridge.connect(IoBridge.java:114)
        at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:192)
        at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
        at java.net.Socket.startupSocket(Socket.java:567)
        at java.net.Socket.<init>(Socket.java:226)
        at javax.net.DefaultSocketFactory.createSocket(DefaultSocketFactory.java:51)
        at io.grpc.okhttp.OkHttpClientTransport$4.run(OkHttpClientTransport.java:541)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExec

```

### PR DESCRIPTION
**Pull request by thebrianchen**
Fixes #904. Verified on Emulator in API 19.

The original check was added in #508 to crash when SSL Ciphers were not loaded.

### CODE REVIEW

**Comment by thebrianchen** [comment]
+CC @var-const

**Comment by MattSkala** [comment]
Don't you think it would make sense to also make the check for `SSLHandshakeException` more specific? Even though we do have Conscrypt bundled with the app, we still occasionally see some crashes caused by `SSLHandshakeException` on Android 4.x.

Here is an example:

```
Fatal Exception: java.lang.IllegalStateException: The Cloud Firestore client failed to establish a secure connection. This is likely a problem with your app, rather than with Cloud Firestore itself. See https://bit.ly/2XFpdma for instructions on how to enable TLS on Android 4.x devices.
       at com.google.firebase.firestore.remote.AbstractStream.close(AbstractStream.java:278)
       at com.google.firebase.firestore.remote.AbstractStream.handleServerClose(AbstractStream.java:387)
       at com.google.firebase.firestore.remote.AbstractStream$StreamObserver.lambda$onClose$3(AbstractStream.java:151)
       at com.google.firebase.firestore.remote.AbstractStream$StreamObserver$$Lambda$4.run(AbstractStream.java)
       at com.google.firebase.firestore.remote.AbstractStream$CloseGuardedRunner.run(AbstractStream.java:67)
       at com.google.firebase.firestore.remote.AbstractStream$StreamObserver.onClose(AbstractStream.java:137)
       at com.google.firebase.firestore.remote.FirestoreChannel$1.onClose(FirestoreChannel.java:131)
       at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
       at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
       at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
       at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:700)
       at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
       at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
       at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
       at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:399)
       at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:500)
       at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:65)
       at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:592)
       at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$700(ClientCallImpl.java:508)
       at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:632)
       at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
       at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:152)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:265)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor$DelayedStartFactory.run(AsyncQueue.java:219)
       at java.lang.Thread.run(Thread.java:841)

Caused by javax.net.ssl.SSLHandshakeException: com.android.org.bouncycastle.jce.exception.ExtCertPathValidatorException: Could not validate certificate: Certificate expired at Tue Dec 10 15:25:38 EET 2019 (compared to Thu Dec 12 17:00:32 EET 2019)
       at com.google.android.gms.org.conscrypt.ConscryptFileDescriptorSocket.startHandshake(ConscryptFileDescriptorSocket.java:42)
       at com.google.android.gms.org.conscrypt.KitKatPlatformOpenSSLSocketImplAdapter.startHandshake(KitKatPlatformOpenSSLSocketImplAdapter.java)
       at io.grpc.okhttp.OkHttpProtocolNegotiator.negotiate(OkHttpProtocolNegotiator.java:92)
       at io.grpc.okhttp.OkHttpProtocolNegotiator$AndroidNegotiator.negotiate(OkHttpProtocolNegotiator.java:147)
       at io.grpc.okhttp.OkHttpTlsUpgrader.upgrade(OkHttpTlsUpgrader.java:63)
       at io.grpc.okhttp.OkHttpClientTransport$4.run(OkHttpClientTransport.java:557)
       at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:841)

Caused by java.security.cert.CertificateException: com.android.org.bouncycastle.jce.exception.ExtCertPathValidatorException: Could not validate certificate: Certificate expired at Tue Dec 10 15:25:38 EET 2019 (compared to Thu Dec 12 17:00:32 EET 2019)
       at com.android.org.conscrypt.TrustManagerImpl.checkTrusted(TrustManagerImpl.java:308)
       at com.android.org.conscrypt.TrustManagerImpl.checkServerTrusted(TrustManagerImpl.java:202)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.google.android.gms.org.conscrypt.Platform.checkTrusted(Platform.java:2)
       at com.google.android.gms.org.conscrypt.Platform.checkServerTrusted(Platform.java:3)
       at com.google.android.gms.org.conscrypt.ConscryptFileDescriptorSocket.verifyCertificateChain(ConscryptFileDescriptorSocket.java:12)
       at com.google.android.gms.org.conscrypt.NativeCrypto.SSL_do_handshake(NativeCrypto.java)
       at com.google.android.gms.org.conscrypt.NativeSsl.doHandshake(NativeSsl.java:7)
       at com.google.android.gms.org.conscrypt.ConscryptFileDescriptorSocket.startHandshake(ConscryptFileDescriptorSocket.java:14)
       at com.google.android.gms.org.conscrypt.KitKatPlatformOpenSSLSocketImplAdapter.startHandshake(KitKatPlatformOpenSSLSocketImplAdapter.java)
       at io.grpc.okhttp.OkHttpProtocolNegotiator.negotiate(OkHttpProtocolNegotiator.java:92)
       at io.grpc.okhttp.OkHttpProtocolNegotiator$AndroidNegotiator.negotiate(OkHttpProtocolNegotiator.java:147)
       at io.grpc.okhttp.OkHttpTlsUpgrader.upgrade(OkHttpTlsUpgrader.java:63)
       at io.grpc.okhttp.OkHttpClientTransport$4.run(OkHttpClientTransport.java:557)
       at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:841)

Caused by com.android.org.bouncycastle.jce.exception.ExtCertPathValidatorException: Could not validate certificate: Certificate expired at Tue Dec 10 15:25:38 EET 2019 (compared to Thu Dec 12 17:00:32 EET 2019)
       at com.android.org.bouncycastle.jce.provider.RFC3280CertPathUtilities.processCertA(RFC3280CertPathUtilities.java:1488)
       at com.android.org.bouncycastle.jce.provider.PKIXCertPathValidatorSpi.engineValidate(PKIXCertPathValidatorSpi.java:305)
       at java.security.cert.CertPathValidator.validate(CertPathValidator.java:190)
       at com.android.org.conscrypt.TrustManagerImpl.checkTrusted(TrustManagerImpl.java:295)
       at com.android.org.conscrypt.TrustManagerImpl.checkServerTrusted(TrustManagerImpl.java:202)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.google.android.gms.org.conscrypt.Platform.checkTrusted(Platform.java:2)
       at com.google.android.gms.org.conscrypt.Platform.checkServerTrusted(Platform.java:3)
       at com.google.android.gms.org.conscrypt.ConscryptFileDescriptorSocket.verifyCertificateChain(ConscryptFileDescriptorSocket.java:12)
       at com.google.android.gms.org.conscrypt.NativeCrypto.SSL_do_handshake(NativeCrypto.java)
       at com.google.android.gms.org.conscrypt.NativeSsl.doHandshake(NativeSsl.java:7)
       at com.google.android.gms.org.conscrypt.ConscryptFileDescriptorSocket.startHandshake(ConscryptFileDescriptorSocket.java:14)
       at com.google.android.gms.org.conscrypt.KitKatPlatformOpenSSLSocketImplAdapter.startHandshake(KitKatPlatformOpenSSLSocketImplAdapter.java)
       at io.grpc.okhttp.OkHttpProtocolNegotiator.negotiate(OkHttpProtocolNegotiator.java:92)
       at io.grpc.okhttp.OkHttpProtocolNegotiator$AndroidNegotiator.negotiate(OkHttpProtocolNegotiator.java:147)
       at io.grpc.okhttp.OkHttpTlsUpgrader.upgrade(OkHttpTlsUpgrader.java:63)
       at io.grpc.okhttp.OkHttpClientTransport$4.run(OkHttpClientTransport.java:557)
       at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:841)

Caused by java.security.cert.CertificateExpiredException: Certificate expired at Tue Dec 10 15:25:38 EET 2019 (compared to Wed Dec 11 11:18:18 EET 2019)
       at com.google.android.gms.org.conscrypt.OpenSSLX509Certificate.checkValidity(OpenSSLX509Certificate.java:3)
       at com.android.org.bouncycastle.jce.provider.RFC3280CertPathUtilities.processCertA(RFC3280CertPathUtilities.java:1483)
       at com.android.org.bouncycastle.jce.provider.PKIXCertPathValidatorSpi.engineValidate(PKIXCertPathValidatorSpi.java:305)
       at java.security.cert.CertPathValidator.validate(CertPathValidator.java:190)
       at com.android.org.conscrypt.TrustManagerImpl.checkTrusted(TrustManagerImpl.java:295)
       at com.android.org.conscrypt.TrustManagerImpl.checkServerTrusted(TrustManagerImpl.java:202)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.google.android.gms.org.conscrypt.Platform.checkTrusted(Platform.java:2)
       at com.google.android.gms.org.conscrypt.Platform.checkServerTrusted(Platform.java:3)
       at com.google.android.gms.org.conscrypt.ConscryptFileDescriptorSocket.verifyCertificateChain(ConscryptFileDescriptorSocket.java:12)
       at com.google.android.gms.org.conscrypt.NativeCrypto.SSL_do_handshake(NativeCrypto.java)
       at com.google.android.gms.org.conscrypt.NativeSsl.doHandshake(NativeSsl.java:7)
       at com.google.android.gms.org.conscrypt.ConscryptFileDescriptorSocket.startHandshake(ConscryptFileDescriptorSocket.java:14)
       at com.google.android.gms.org.conscrypt.KitKatPlatformOpenSSLSocketImplAdapter.startHandshake(KitKatPlatformOpenSSLSocketImplAdapter.java)
       at io.grpc.okhttp.OkHttpProtocolNegotiator.negotiate(OkHttpProtocolNegotiator.java:92)
       at io.grpc.okhttp.OkHttpProtocolNegotiator$AndroidNegotiator.negotiate(OkHttpProtocolNegotiator.java:147)
       at io.grpc.okhttp.OkHttpTlsUpgrader.upgrade(OkHttpTlsUpgrader.java:63)
       at io.grpc.okhttp.OkHttpClientTransport$4.run(OkHttpClientTransport.java:557)
       at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:841)
```

This exception  has been caused by an incorrect system time on the client device, but we can't handle such case in any way, as Firestore SDK crashes the main thread.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/Datastore.java [hunk 1]
```diff
@@ -241,14 +240,20 @@ public static boolean isPermanentError(FirebaseFirestoreException.Code code) {
    *
    * <p>This error is non-recoverable and must be addressed by the app developer.
    */
-  public static boolean isSslHandshakeError(Status status) {
+  public static boolean isMissingSslCiphers(Status status) {
     Status.Code code = status.getCode();
     Throwable t = status.getCause();
 
+    // Check for the presence of a cipher error in the even of an SSLHandshakeException. This is
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/Datastore.java]: s/even/event

**Pull request merged**